### PR TITLE
Add `GeoDatasetDriver` `GeoDatasetSource` and new `GeoDatasetAdapter`

### LINKED ISSUE #895: new `GeodatasetDriver` and `GeodatasetAdapter` classes
### Kind of request

Changing existing functionality

### Enhancement Description

transform https://github.com/Deltares/hydromt/blob/main/hydromt/data_adapter/geodataset.py into the new data source / driver / adapter structure

### Use case

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by savente93**
## Issue addressed
Fixes #895 

## Explanation
This PR adds the `GeoDatasetSource`, `GeoDatasetDriver`, `VectorGeoDatasetDriver`, and addapts the `GeoDatasetAdapter` to the new architecture. From there it's all fairly business as usual.

I've also added property to the drivers to flag whether the drivers can write or not so the users can get a nicer error without having to do things like read in the data if the driver does not support it. Happy to look for a different mechanism if necessary but I do think this is good functionality to have. Added tests where possible, mostly copied from RasterDatasetDriver, so if I forgot to update a docstring somewhere feel free to point it out. 

## Checklist
- [x] Updated tests or added new tests
- [x] Branch is up to date with `main`
- [x] Tests & pre-commit hooks pass
- [ ] Updated documentation if needed
- [x] Updated changelog.rst if needed
- [x] For predefined catalogs: update the catalog version in the file itself, the references in data/predefined_catalogs.yml, and the changelog in data/changelog.rst

## Additional Notes (optional)
Add any additional notes or information that may be helpful.

### CODE REVIEW

**Comment by savente93** [comment]
I only just realized that I forgot to add tests for the new `GeoDatasetAdaper`. I'm doing that now

**Review by Jaapel — CHANGES_REQUESTED** [review]
Most of the comments are on the simplification using the `mask` argument.
Thanks for the hard work!

File: hydromt/data_source/geodataset.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+        if ds is None:
+            ds = self.read_data()
+        return (
+            ds[ds.raster.time_dim].min().values,
+            ds[ds.raster.time_dim].max().values,
+        )
+
+    def to_stac_catalog(
+        self,
+        on_error: ErrorHandleMethod = ErrorHandleMethod.COERCE,
```
[the remark this answers is not in the document]
Comment (savente93) [reply hydromt/data_source/geodataset.py:263]: very true, though I think that we can best do that in #802

File: hydromt/data_adapter/utils.py [hunk 2]
```diff
@@ -124,3 +59,82 @@ def has_no_data(
...
+        else:
+            nodata = metadata.nodata
+        for k in ds.data_vars:
+            mv = nodata.get(k, None)
+            if hasattr(ds[k], "vector"):
+                if mv is not None and ds[k].vector.nodata is None:
+                    ds[k].vector.set_nodata(mv)
+            if hasattr(ds[k], "raster"):
+                if mv is not None and ds[k].raster.nodata is None:
+                    ds[k].raster.set_nodata(mv)
```
[the remark this answers is not in the document]
Comment (savente93) [reply hydromt/data_adapter/utils.py]: I think we should but I talked briefly to  Hélène about this and apparently this is is non-trivial so I left it out fo scope for this one. I do believe there is a lot more generic things we can do if we enable that. If you agree to keep it out of scope for this one we should make a ticket.

[the remark this answers is not in the document]
Comment (Jaapel) [reply hydromt/data_adapter/utils.py]: agreed! https://github.com/Deltares/hydromt/issues/915

**Comment by savente93** [comment]
@Jaapel  I fixed all your comments. It looks okay but whether everything works will have to be seen during integration tests. I think it should be all good now. I looked at the sonar cloud failures and they are all things that are outside the scope of this PR like the TODO mentioning the no data, so I think this is okay for now. Can you take another look?

**Review by Jaapel — APPROVED** [review]

File: hydromt/data_source/geodataset.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+        driver.write(
+            file_path,
+            ds,
+            **kwargs,
+        )
+
+        return self.model_copy(update=update)
+
+    def get_bbox(self, crs: Optional[CRS], detect: bool = True) -> TotalBounds:
```
Comment (Jaapel) [inline hydromt/data_source/geodataset.py:142]:
```suggestion
    def get_bbox(self, crs: Optional[CRS] = None, detect: bool = True) -> TotalBounds:
```

**Pull request merged**